org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/ExecutorServiceProvider.java

```diff
@@ -24,17 +24,18 @@
  * Provider for executor services. By calling {@link #dispose()} all created executor services are shut down.
  * <p>
  * In some situations it is necessary to use multiple instances of executor services in order to avoid deadlocks. That
- * can be achieved with the {@link #get(String)}�method, which will return a different instance for each key.
+ * can be achieved with the {@link #get(String)}�method, which will return a different instance for each key. </p>
  */
 @Singleton
 public class ExecutorServiceProvider implements Provider<ExecutorService>, IDisposable {
+	
+	private final Map<String, ExecutorService> instanceCache = new ConcurrentHashMap<>(3);
```

Comment:
Given that ConcurrentHashMap does not allow null-keys or null-values, it may also be ok to use Collections.synchronizedMap here, since I don't expect a lot of pressure on the lookup of an ExecutorService.

`get(String key)` should be rewritten to use `computeIfAbsent(..)`.

Reply:
`dispose` must sync on the map, too.

Reply:
> get(String key) should be rewritten to use `computeIfAbsent(..)`.

I can think of two ways; however, I am not convinced by either. In the first one, calling  `computeIfAbsent(...)` inside the `if (result == null)` doesn't make much sense. We already know that the value is `null` before calling `computeIfAbsent(...)`. The second one would block threads even if the map has a non-null value for the key requested by the second thread.  
@szarnekow Please suggest if there is any other way to rewrite this method.  

```
ExecutorService result = instanceCache.get(key);                            
if (result == null) {                                                       
	synchronized (instanceCache) {                                          
		result = instanceCache.computeIfAbsent(key, k -> createInstance(k));
	}                                                                       
}                                                                           
return result; 
```
```
synchronized (instanceCache) {                                         
	return instanceCache.computeIfAbsent(key, k -> createInstance(k)); 
} 
```

Reply:
Steps:
1. Use a `Collections.synchronizedMap(new HashMap<>())` for the field initializer
2. Rewrite and simplify get(String) towards `public ExecutorService get(String key) { return instanceCache.computeIfAbsent(key, this::createInstance)); }`
3. fix dispose to sync on the instanceCache, too: `public void dipose() { synchronized(instanceCache) { .. } }`

Reply:
@szarnekow `computeIfAbsent(...)` is not atomic. Don't we have to sync on the map?

Reply:
`SynchronizedMap.computeIfAbsent(K, Function<? super K, ? extends V>)` is safe.

Reply:
ok, tx for clarification.